src/PriceCalculatorKata/Product.cs

```diff
@@ -34,25 +40,39 @@ public double UniversalDiscountPercentage
                 _universalDiscountPercentage = value / 100.0;
             }
         }
-
+        public DiscountPrecedence UniversalDiscountPrecedence { get; set; } = DiscountPrecedence.AfterTax;
         public void CheckPercentageValidation(double value)
         {
             if (value < 0 && value > 100)
                 throw new Exception("Percentage is not valid");
         }
         public void ReportProductPrice()
         {
-            double DiscountAmount = GetTotalDiscountAmount();
-            double TaxAmount = Price * TaxPercentage;
-            double PriceWithTaxAndDiscount = Math.Round(Price + TaxAmount - DiscountAmount, 2);
+            double TotalDiscountAmount = GetUniversalDiscountAmount() + GetUPCDiscountAmount();
+            double TaxAmount = GetTaxAmount(Price);
+            double PriceWithTaxAndDiscount = Math.Round(Price + TaxAmount - TotalDiscountAmount, 2);
             Console.WriteLine($"${Price} before tax and discount and ${PriceWithTaxAndDiscount} after {TaxPercentage * 100}% tax and {UniversalDiscountPercentage * 100}% discount");
-            if (DiscountAmount > 0)
-                Console.WriteLine($"Total Discount amount = {DiscountAmount}");
+            if (TotalDiscountAmount > 0)
+                Console.WriteLine($"Total Discount amount = {TotalDiscountAmount}");
         }
-        private double GetTotalDiscountAmount()
+        private double GetTaxAmount(double Price)
+        {
+            if (UPCDiscountPrecedence == DiscountPrecedence.BeforeTax)
+                Price -= GetUPCDiscountAmount();
+            if (UniversalDiscountPrecedence == DiscountPrecedence.BeforeTax)
+                Price -= GetUniversalDiscountAmount();
+            double TaxAmount = Math.Round(Price * TaxPercentage, 2);
+            Console.WriteLine($"Tax amount = {TaxAmount}");
```

Comment:
Usually these calculation methods should have the business logic to calculate a certain price or whatever, and return that value. No need to print out this to the console. If you want to do that, then you need to move this out to Program.cs maybe, wherever you call this GetTaxAmount.

We should make sure to separate business logic into small units and make them loosely coupled as much as possible.